src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java

```diff
@@ -517,33 +517,37 @@ private void stopConnection(String errorMessage) {
         RuntimeException exception = null;
         this.state.lock();
         try {
+            ConnectionState connectionState = this.state.getConnectionStateUnsynchronized(false);
```

Comment:
If the connection state is null, this will throw. We have run into this, I'm wondering if we can be more defensive and allow null here?

Reply:
Thanks for the feedback. I'm trying to understand how this is happening. stopConnection is only ever called when the transport fires its closed handlers. You can see where it's wired up here:

https://github.com/dotnet/aspnetcore/blob/88bf14399bb6c3557a73ce1a38ba7d7e9f27ac88/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java#L276-L279

At the point stopConnection is wired up as the transport's close handler, the ConnectionState is already in scope. That means we can fix any issues with a null ConnectionState here by simply capturing the ConnectionState in the setOnClose lambda and passing the ConnectionState as an argument to stopConnection. That way there's absolutely no risk of referencing the wrong ConnectionState or it being null. You could then verify the passed-in ConnectionState matches what's returned from getConnectionStateUnsynchronized() before calling changeState(HubConnectionState.DISCONNECTED) or mutating anything else outside of the ConnectionState object.

https://github.com/dotnet/aspnetcore/blob/88bf14399bb6c3557a73ce1a38ba7d7e9f27ac88/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java#L1482-L1487

That said, @BrennanConroy @aaronoe Do either of you have any ideas about how ConnectionState could be null here today? As far as I can tell, this is the only method that currently sets ConnectionState to null, and it cannot be called before it's hooked to the transport's close event which happens after the ConnectionState is set.

There are two possibilities I can think of at the moment:

1. The transport triggering stopConnection multiple times for a single connection.
2. The connection starting and stopping twice. If the first call to stopConnection is delayed long enough, the ConnectionState might have been overwritten by the second connection before the first call to stopConnection fires.

I'm thinking instead of just checking the HubConnectionState in start() we should also verify ConnectionState is null before continuing.

https://github.com/dotnet/aspnetcore/blob/88bf14399bb6c3557a73ce1a38ba7d7e9f27ac88/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/HubConnection.java#L224-L234

Looking at all this also has me wondering what happens if transport.start() fails so the transport never fires stopConnection. Is the ConnectionState still cleared? It appears not.

Reply:
> Do either of you have any ideas about how ConnectionState could be null here today?

Yes, if there is an exception during the websockets `onClosing` method it will call `onFailure`. This is why I have wrapped the user `onClosed` callbacks, and modified the changing state in `stopConnection` so that both vectors will no longer create exceptions.

> Looking at all this also has me wondering what happens if transport.start() fails so the transport never fires stopConnection. Is the ConnectionState still cleared? It appears not.

Correct, but the state will be disconnected and the next start will overwrite the ConnectionState.

Reply:
For reference, these are the stacktraces for the two separate (but similar) issues:
```
java.lang.RuntimeException: The HubConnection failed to transition from the 'CONNECTED' state to the 'DISCONNECTED' state because it was actually in the 'CONNECTING' state.
  at com.microsoft.signalr.HubConnection$ReconnectingConnectionState.changeState(HubConnection.java:1512)
  at com.microsoft.signalr.HubConnection.stopConnection(HubConnection.java:538)
  at com.microsoft.signalr.HubConnection.lambda$start$6(HubConnection.java:279)
  at com.microsoft.signalr.-$$Lambda$HubConnection$plWe2n559m5117SGBt0gopoPs3g.invoke(-.java:2)
  at com.microsoft.signalr.WebSocketTransport.onClose(WebSocketTransport.java:92)
  at com.microsoft.signalr.WebSocketTransport.lambda$start$1(WebSocketTransport.java:55)
  at com.microsoft.signalr.-$$Lambda$WebSocketTransport$vlt_DXG3u2t5OJvAYlxBdjX-Br4.invoke(-.java:2)
  at com.microsoft.signalr.OkHttpWebSocketWrapper$SignalRWebSocketListener.onFailure(OkHttpWebSocketWrapper.java:124)
  at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:592)
  at okhttp3.internal.ws.RealWebSocket$connect$1.onResponse(RealWebSocket.java:174)
  at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.java:519)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:919)
```

and 

```
java.lang.RuntimeException: Connection is not active.
  at com.microsoft.signalr.HubConnection$ReconnectingConnectionState.getConnectionStateUnsynchronized(HubConnection.java:1484)
  at com.microsoft.signalr.HubConnection.stopConnection(HubConnection.java:522)
  at com.microsoft.signalr.HubConnection.lambda$start$6(HubConnection.java:279)
  at com.microsoft.signalr.-$$Lambda$HubConnection$plWe2n559m5117SGBt0gopoPs3g.invoke(-.java:2)
  at com.microsoft.signalr.WebSocketTransport.onClose(WebSocketTransport.java:90)
  at com.microsoft.signalr.WebSocketTransport.lambda$start$1(WebSocketTransport.java:55)
  at com.microsoft.signalr.-$$Lambda$WebSocketTransport$vlt_DXG3u2t5OJvAYlxBdjX-Br4.invoke(-.java:2)
  at com.microsoft.signalr.OkHttpWebSocketWrapper$SignalRWebSocketListener.onFailure(OkHttpWebSocketWrapper.java:148)
  at okhttp3.internal.ws.RealWebSocket.failWebSocket(RealWebSocket.java:592)
  at okhttp3.internal.ws.RealWebSocket$WriterTask.runOnce(RealWebSocket.java:622)
  at okhttp3.internal.concurrent.TaskRunner.runTask(TaskRunner.java:116)
  at okhttp3.internal.concurrent.TaskRunner.access$getLogger$cp(TaskRunner.java:42)
  at okhttp3.internal.concurrent.TaskRunner.access$runTask(TaskRunner.java:42)
  at okhttp3.internal.concurrent.TaskRunner$runnable$1.run(TaskRunner.java:65)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
  at java.lang.Thread.run(Thread.java:919)

```

I am also wondering if this is the right websocket callback being used. Looking at the docs, there is both [onClosed](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-web-socket-listener/on-closed/) and [onClosing](https://square.github.io/okhttp/4.x/okhttp/okhttp3/-web-socket-listener/on-closing/). onClosed specifically calls out that only one call is made to that method, but onClosing does not call that out specifically. Could you confirm that the correct callback is being used here:
https://github.com/dotnet/aspnetcore/blob/5bac7b01e1fcc2af4f495a2e948f71798ffafb8c/src/SignalR/clients/java/signalr/core/src/main/java/com/microsoft/signalr/OkHttpWebSocketWrapper.java#L104

Reply:
> I am also wondering if this is the right websocket callback being used. 

Interesting point about onClosing vs onClosed. onClosed does sound more correct. In any case, I think we should add a check in stopConnection that logs a critical error if it's called more than once to be sure. It could be that there's yet something else we're not aware of throws and triggers onFailure.

> Correct, but the state will be disconnected and the next start will overwrite the ConnectionState.

We should either always clear it when the connection becomes disconnected or always leave it. I think clearing it makes more sense.

> That code only runs if transport.start results in an error, in which case stopConnection should never get called.

Wouldn't the handshake failing or timing out also cause that code to run? I remember this being an issue before. Was this fixed? Or am I mistaken?